### PR TITLE
[INFRA/CORE] Refactor common folder + actions

### PR DESCRIPTION
**Pull request by lhl2617**
# Summary

- `common` package may end up too messy, so refactor it to contain packages. Also include a dependency map to prevent import cycles.
- Change `Action` interface and dispatch style.

## Test Plan
- Added a fake action to get `Team1` to give 10 resources to `Team2`:
```
2020/12/18 15:23:21 [GAMESTATE]: start DispatchActions
2020/12/18 15:23:21 [GAMESTATE]: Team1 gave 10 resources to Team2
2020/12/18 15:23:21 [GAMESTATE]: finish DispatchActions
```

- `go test ./...`
```
?       github.com/SOMAS2020/SOMAS2020  [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team1   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team2   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team3   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team4   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team5   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team6   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/common/baseclient       [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/common/config   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/common/gamestate        [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/common/rules    (cached)
?       github.com/SOMAS2020/SOMAS2020/internal/common/shared   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/server  0.234s
?       github.com/SOMAS2020/SOMAS2020/pkg/testutils    [no test files]
```

- `go run .`
```
2020/12/18 14:57:16 [SERVER]: start runTurn
2020/12/18 14:57:16 [SERVER]: TURN: 1, Season: 1
2020/12/18 14:57:16 [SERVER]: start updateIsland
2020/12/18 14:57:16 [Team6]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 14:57:16 [Team1]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 14:57:16 [Team2]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 14:57:16 [Team3]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 14:57:16 [Team4]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 14:57:16 [Team5]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 14:57:16 [SERVER]: finish updateIsland
2020/12/18 14:57:16 [SERVER]: start runOrgs
2020/12/18 14:57:16 [SERVER]: start runIIGO
2020/12/18 14:57:16 [SERVER]: finish runIIGO
2020/12/18 14:57:16 [SERVER]: start runIIFO
2020/12/18 14:57:16 [SERVER]: finish runIIFO
2020/12/18 14:57:16 [SERVER]: start runIITO
2020/12/18 14:57:16 [SERVER]: finish runIITO
2020/12/18 14:57:16 [SERVER]: finish runOrgs
2020/12/18 14:57:16 [SERVER]: start getAndDispatchEndOfTurnActions
2020/12/18 14:57:16 [Team6]: EndOfTurnActions
2020/12/18 14:57:16 [Team1]: EndOfTurnActions
2020/12/18 14:57:16 [Team2]: EndOfTurnActions
2020/12/18 14:57:16 [Team3]: EndOfTurnActions
2020/12/18 14:57:16 [Team4]: EndOfTurnActions
2020/12/18 14:57:16 [Team5]: EndOfTurnActions
2020/12/18 14:57:16 [GAMESTATE]: start DispatchActions
2020/12/18 14:57:16 [GAMESTATE]: finish DispatchActions
2020/12/18 14:57:16 [SERVER]: finish getAndDispatchEndOfTurnActions
2020/12/18 14:57:16 [SERVER]: start endOfTurn
2020/12/18 14:57:16 [SERVER]: start probeDisaster
2020/12/18 14:57:16 [SERVER]: finish probeDisaster
2020/12/18 14:57:16 [SERVER]: start deductCostOfLiving
2020/12/18 14:57:16 [SERVER]: finish deductCostOfLiving
2020/12/18 14:57:16 [SERVER]: start updateIslandLivingStatus
2020/12/18 14:57:16 [SERVER]: finish updateIslandLivingStatus
2020/12/18 14:57:16 [SERVER]: finish endOfTurn
2020/12/18 14:57:16 [SERVER]: finish runTurn
2020/12/18 14:57:16 [SERVER]: start runTurn
2020/12/18 14:57:16 [SERVER]: TURN: 2, Season: 1
2020/12/18 14:57:16 [SERVER]: start updateIsland
2020/12/18 14:57:16 [Team1]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 14:57:16 [Team2]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 14:57:16 [Team3]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 14:57:16 [Team4]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 14:57:16 [Team5]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 14:57:16 [Team6]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 14:57:16 [SERVER]: finish updateIsland
2020/12/18 14:57:16 [SERVER]: start runOrgs
2020/12/18 14:57:16 [SERVER]: start runIIGO
2020/12/18 14:57:16 [SERVER]: finish runIIGO
2020/12/18 14:57:16 [SERVER]: start runIIFO
2020/12/18 14:57:16 [SERVER]: finish runIIFO
2020/12/18 14:57:16 [SERVER]: start runIITO
2020/12/18 14:57:16 [SERVER]: finish runIITO
2020/12/18 14:57:16 [SERVER]: finish runOrgs
2020/12/18 14:57:16 [SERVER]: start getAndDispatchEndOfTurnActions
2020/12/18 14:57:16 [Team5]: EndOfTurnActions
2020/12/18 14:57:16 [Team6]: EndOfTurnActions
2020/12/18 14:57:16 [Team1]: EndOfTurnActions
2020/12/18 14:57:16 [Team2]: EndOfTurnActions
2020/12/18 14:57:16 [Team3]: EndOfTurnActions
2020/12/18 14:57:16 [Team4]: EndOfTurnActions
2020/12/18 14:57:16 [GAMESTATE]: start DispatchActions
2020/12/18 14:57:16 [GAMESTATE]: finish DispatchActions
2020/12/18 14:57:16 [SERVER]: finish getAndDispatchEndOfTurnActions
2020/12/18 14:57:16 [SERVER]: start endOfTurn
2020/12/18 14:57:16 [SERVER]: start probeDisaster
2020/12/18 14:57:16 [SERVER]: finish probeDisaster
2020/12/18 14:57:16 [SERVER]: start deductCostOfLiving
2020/12/18 14:57:16 [SERVER]: finish deductCostOfLiving
2020/12/18 14:57:16 [SERVER]: start updateIslandLivingStatus
2020/12/18 14:57:16 [SERVER]: finish updateIslandLivingStatus
2020/12/18 14:57:16 [SERVER]: finish endOfTurn
2020/12/18 14:57:16 [SERVER]: finish runTurn
2020/12/18 14:57:16 [SERVER]: Max turns '2' reached or exceeded
===== START OF TURN 1 (END OF TURN 0) =====
gamestate.GameState{Season:0x1, Turn:0x1, ClientInfos:map[shared.ClientID]gamestate.ClientInfo{Team1:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:gamestate.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 2 (END OF TURN 1) =====
gamestate.GameState{Season:0x1, Turn:0x2, ClientInfos:map[shared.ClientID]gamestate.ClientInfo{Team1:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:gamestate.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 3 (END OF TURN 2) =====
gamestate.GameState{Season:0x1, Turn:0x3, ClientInfos:map[shared.ClientID]gamestate.ClientInfo{Team1:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:gamestate.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
```

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

File: internal/common/baseclient/baseclient.go [hunk 1]
```diff
@@ -1,12 +1,31 @@
-package common
+// Package baseclient contains the Client interface as well as a base client implementation.
```
Comment (lhl2617) [inline internal/common/baseclient/baseclient.go:1]: moved from `internal/common/client.go`

File: internal/common/baseclient/helper.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package baseclient
```
Comment (lhl2617) [inline internal/common/baseclient/helper.go:1]: moved from `internal/common/client.go`

File: internal/common/gamestate/doc.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/*
+Package gamestate contains information about the current game state.
+It also contains action code that mutates gameState.
+
```
Comment (preetl) [inline internal/common/gamestate/doc.go:4]: Really helpful!

**Review by Darrekt — APPROVED** [review]

**Pull request merged**